src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp

```diff
@@ -15,751 +15,879 @@
 #include "InstanceScript.h"
 #include "MotionMaster.h"
 #include "ObjectAccessor.h"
 #include "ScriptedCreature.h"
 #include "SpellAuras.h"
 #include "SpellMgr.h"
 #include "SpellScript.h"
+#include "Spell.h"
 #include "TemporarySummon.h"
 #include "ulduar.h"
 
-enum AssemblySpells
+enum Spells
 {
     // General
-    SPELL_SUPERCHARGE                            = 61920,
-    SPELL_BERSERK                                = 47008, // Hard enrage, don't know the correct ID.
-    SPELL_KILL_CREDIT                            = 65195, // spell_dbc
+    SPELL_SUPERCHARGE                           = 61920,
+    SPELL_BERSERK                               = 62535,
+    SPELL_KILL_CREDIT                           = 65195, // Serverside spell
 
     // Steelbreaker
-    SPELL_HIGH_VOLTAGE                           = 61890,
-    SPELL_FUSION_PUNCH                           = 61903,
-    SPELL_STATIC_DISRUPTION                      = 44008,
-    SPELL_OVERWHELMING_POWER                     = 64637,
-    SPELL_ELECTRICAL_CHARGE                      = 61902,
+    SPELL_HIGH_VOLTAGE                          = 61890,
+    SPELL_STATIC_DISRUPTION                     = 44008,
+    SPELL_ELECTRICAL_CHARGE                     = 61902,
 
     // Runemaster Molgeim
-    SPELL_SHIELD_OF_RUNES                        = 62274,
-    SPELL_SHIELD_OF_RUNES_BUFF                   = 62277,
-    SPELL_SUMMON_RUNE_OF_POWER                   = 63513,
-    SPELL_RUNE_OF_DEATH                          = 62269,
-    SPELL_RUNE_OF_SUMMONING                      = 62273, // This is the spell that summons the rune
-    SPELL_RUNE_OF_SUMMONING_SUMMON               = 62020, // Spell that summons
+    SPELL_RUNE_OF_POWER_PRE_FIGHT               = 61975,
+    SPELL_RUNE_OF_POWER                         = 61973,
+    SPELL_SHIELD_OF_RUNES                       = 62274,
+    SPELL_SHIELD_OF_RUNES_BUFF                  = 62277,
+    SPELL_RUNE_OF_DEATH                         = 62269,
+    SPELL_RUNE_OF_SUMMONING                     = 62273,
+    SPELL_RUNE_OF_SUMMONING_SUMMON              = 62020,
 
     // Stormcaller Brundir
-    SPELL_CHAIN_LIGHTNING                        = 61879,
-    SPELL_OVERLOAD                               = 61869,
-    SPELL_LIGHTNING_WHIRL                        = 61915,
-    SPELL_LIGHTNING_TENDRILS                     = 61887,
-    SPELL_LIGHTNING_TENDRILS_VISUAL              = 61883,
-    SPELL_STORMSHIELD                            = 64187
+    SPELL_LIGHTNING_CHANNEL_PRE_FIGHT_VISUAL    = 61942,
+    SPELL_CHAIN_LIGHTNING                       = 61879,
+    SPELL_LIGHTNING_WHIRL                       = 61915,
+    SPELL_LIGHTNING_TENDRILS_VISUAL             = 61883,
+    SPELL_STORMSHIELD                           = 64187,
+    SPELL_RANDOM_AGGRO_PERIODIC                 = 61906
 };
 
-enum AssemblyEvents
-{
-    // General
-    EVENT_BERSERK                                = 1,
+#define SPELL_OVERLOAD              RAID_MODE<uint32>(61869, 63481)
```

Comment:
why not use spelldifficulty_dbc to handle with it?
make port ez for masters/4.x?

Reply:
Read further below. It's to avoid spellmgr lookups when it's not needed.

Reply:
ummm if these are introduced in the table, the lookups should not be too complex, IIRC everything is stored in an unordered map.

Whether thats theoretically more or less efficient falls off the scope of a script. Scripts should have guidelines, scripts should use what already exists and is supported.

You might argue, but this is used in many other places, and that is correct, but new scripts or old script rewrites dont tend to, I might be wrong though, but I dont recall the opposite.

And we have @Keader and other folks to thank for that. Consistency is good, standards are good, lets keep going towards that direction.

Reply:
Agreed. "I save a single hashmap lookup every 15 seconds" is the epitome of premature optimization.